classes/key_manager.php

```diff
@@ -0,0 +1,144 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Access key manager.
+ *
+ * @package    tool_pdfpages
+ * @author     Tom Dickman <tomdickman@catalyst-au.net>
+ * @copyright  2021 Catalyst IT
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace tool_pdfpages;
+
+defined('MOODLE_INTERNAL') || die();
+
+/**
+ * Access key manager.
+ *
+ * @package    tool_pdfpages
+ * @author     Tom Dickman <tomdickman@catalyst-au.net>
+ * @copyright  2021 Catalyst IT
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class key_manager {
+
+    /**
+     * Create a user key.
+     *
+     * @param int $instanceid the instance ID to create key for.
+     * @param string $iprestriction optional IP range to restrict access to.
+     *
+     * @return string the created user key value.
+     * @throws \moodle_exception if user doesn't have permission to create key.
+     */
+    public static function create_user_key(int $instanceid, string $iprestriction = ''): string {
+        global $USER;
+
+        require_capability('tool/pdfpages:generatepdf', \context_system::instance());
+
+        $iprestriction = !empty($iprestriction) ? $iprestriction : null;
+
+        self::delete_user_key($instanceid);
+
+        $ttl = get_config('tool_pdfpages', 'accesskeyttl');
+        $expirationtime = !empty($ttl) ? (time() + $ttl) : (time() + MINSECS);
+
+        return create_user_key('tool/pdfpages', $USER->id, $instanceid, $iprestriction, $expirationtime);
+    }
+
+    /**
+     * Delete a user key.
+     *
+     * @param int $instanceid the instance ID to delete user key for.
+     *
+     * @return bool true on success.
+     */
+    public static function delete_user_key(int $instanceid): bool {
+        global $DB, $USER;
+
+        $record = [
+            'script' => 'tool/pdfpages',
+            'userid' => $USER->id,
+            'instance' => $instanceid
+        ];
+
+        return $DB->delete_records('user_private_key', $record);
+    }
+
+    /**
+     * Get a unique access key instance ID for a filename.
+     *
+     * @param string $filename the filename to get instance ID for.
+     *
+     * @return int the instance ID.
+     */
+    public static function get_instance_id(string $filename): int {
+        return (int) substr(base_convert(sha1($filename), 16, 10), 0, 18);
+    }
+
+    /**
+     * Login user with access key.
+     *
+     * @param string $key access key to use for user validation, this is required to login user and allow access of target page.
+     * @param int $instanceid the instance ID of key to login with.
+     */
+    final public static function login_with_key(string $key, int $instanceid) {
```

Comment:
it feels like this method shouldn't be part of that class. I'd expect key_manager to deal with keys and login functionality could be moved to login_manager for example.

Reply:
Moved to a new `login_manager`